cranelift/codegen/src/isa/x64/abi.rs

```diff
@@ -0,0 +1,457 @@
+//! Implementation of the standard x64 ABI.
+
+use alloc::vec::Vec;
+use regalloc::{RealReg, Reg, RegClass, Set, SpillSlot, Writable};
+
+use crate::ir::{self, types, types::*, ArgumentExtension, StackSlot, Type};
+use crate::isa::{self, x64::inst::*};
+use crate::machinst::*;
+use crate::settings;
+
+use args::*;
+
+#[derive(Clone, Debug)]
+enum ABIArg {
+    Reg(RealReg),
+    _Stack,
+}
+
+#[derive(Clone, Debug)]
+enum ABIRet {
+    Reg(RealReg),
+    _Stack,
+}
+
+pub(crate) struct X64ABIBody {
+    args: Vec<ABIArg>,
+    rets: Vec<ABIRet>,
+
+    /// Offsets to each stack slot.
+    _stack_slots: Vec<usize>,
+
+    /// Total stack size of all the stack slots.
+    stack_slots_size: usize,
+
+    /// Clobbered registers, as indicated by regalloc.
+    clobbered: Set<Writable<RealReg>>,
+
+    /// Total number of spill slots, as indicated by regalloc.
+    num_spill_slots: Option<usize>,
+
+    /// Calculated while creating the prologue, and used when creating the epilogue. Amount by
+    /// which RSP is adjusted downwards to allocate the spill area.
+    frame_size_bytes: Option<usize>,
+
+    call_conv: isa::CallConv,
+
+    /// The settings controlling this function's compilation.
+    flags: settings::Flags,
+}
+
+fn in_int_reg(ty: types::Type) -> bool {
+    match ty {
+        types::I8
+        | types::I16
+        | types::I32
+        | types::I64
+        | types::B1
+        | types::B8
+        | types::B16
+        | types::B32
+        | types::B64 => true,
+        _ => false,
+    }
+}
+
+fn get_intreg_for_arg_systemv(idx: usize) -> Option<Reg> {
+    match idx {
+        0 => Some(regs::rdi()),
+        1 => Some(regs::rsi()),
+        2 => Some(regs::rdx()),
+        3 => Some(regs::rcx()),
+        4 => Some(regs::r8()),
+        5 => Some(regs::r9()),
+        _ => None,
+    }
+}
+
+fn get_intreg_for_retval_systemv(idx: usize) -> Option<Reg> {
+    match idx {
+        0 => Some(regs::rax()),
+        1 => Some(regs::rdx()),
+        _ => None,
+    }
+}
+
+fn is_callee_save_systemv(r: RealReg) -> bool {
+    use regs::*;
+    match r.get_class() {
+        RegClass::I64 => match r.get_hw_encoding() as u8 {
+            ENC_RBX | ENC_RBP | ENC_R12 | ENC_R13 | ENC_R14 | ENC_R15 => true,
+            _ => false,
+        },
+        _ => unimplemented!(),
+    }
+}
+
+fn get_callee_saves(regs: Vec<Writable<RealReg>>) -> Vec<Writable<RealReg>> {
+    regs.into_iter()
+        .filter(|r| is_callee_save_systemv(r.to_reg()))
```

Comment:
just making sure I'm following the intended direction: this will eventually have an `if call_conv.extends_windows_fastcall()` arm like the existing x86 abi code, correct?

Reply:
I don't know the details (Benjamin surely will), but no doubt the intent exists to make it possible to support windows-specific ABIs as necessary.

Reply:
Yes, that's the idea! We'd pass the `CallConv`, or we'd select different impl in the caller according to the given `CallConv` here, as done in `x86/abi.rs`.